### ISSUE TITLE
README needs info about 2010 & 2015 VC++ Redistributable, for 64-bit Protoplug

### ISSUE THREAD

**Issue by GezzMC** [pac-dev/protoplug#32]
[title]: # (README needs info about 2010 & 2015 VC++ Redistributable, for 64-bit Protoplug)

Thanks again for the software. I work to keep unused Visual C++ Redistributables off of my PC. Programs install them but don't uninstall them. It's bad when great programs stop working on a new PC.

I got lucky with 32-bit Protoplug, but 64-bit Protoplug needs two versions of Visual C++ Redistributable installed: 2015 for `Lua Protoplug Fx (x64).dll/Lua Protoplug Gen (x64).dll`, and 2010 for `lua51.dll`.

Anyway, the problem is only a documentation problem. I provide links below to the Microsoft web pages I used to get the downloads, and I make a few comments about the file dates of `lua51.dll`.

# Aside Number 1: ReaPlugs/Protoplug as a team, ReaPlugs vs. Protoplug #

Only God knows the future: bad problems will show up fast, bad problems will take time to show up, things will work good enough, or things will work great.

But the plan is like this:
  
* Use [ReaPlugs VST](https://www.reaper.fm/reaplugs/) to run JSFX scripts in other DAWS, 
* so that I can continue to use the [JSFX scripts](https://www.reaper.fm/sdk/js/js.php) I now use (which are scripts I've used to get rid of all normal use of VSTs),
* and then get up to speed with Lua and Protoplug,
* and slowly replace all the JSFX scripts with Lua Protoplug scripts,
* or, worst case, continue to use ReaPlugs and Protoplug together.

It's all good, but comparisons can be made between JSFX/ReaPlugs and Lua/Protoplug. In particular, Lua is a full-blown scripting language, so if it will work for real-time audio processing, it's better.

# Preface comment about Visual Studio 2017 #

As a preface to what download I decided to use, the GitHub Protoplug README says this in section *Compiling from Source*: 

> *"You'll need Visual Studio 2017 (Windows.

I try to use only the minimal dependencies, but I decided to use `Microsoft Visual C++ 2015-2019 Redistributable`. On the page linked to below, it says this:

> Note Visual C++ 2015, 2017 and 2019 all share the same redistributable files.

I know a little more now, but only a little.

# Back to the Download Links #

The file dates of 32-bit and 64-bit `lua51.dll`, of the v1.4.0 downloads, are 2014 file dates. The 32-bit Protoplug works with only `Microsoft Visual C++ 2015 Redistributable (x86)` installed, but 64-bit Protoplug needs two installs of Visuall C++ Redistributable.

In the process of finding the problem, I installed the 2012 and 2013 VC++R, but it was the 2010 version that finally got 64-bit `lua51.dll` working.

For the 64-bit VSTs, I installed `Microsoft Visual C++ 2015-2019 Redistributable (14.21.27702.2)`. For security, I've started to try and make sure I get the latest versions, but Microsoft doesn't work hard to make doing that easy. I finally found this page:

* Page title: _The latest supported Visual C++ downloads_
* https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads
* Scroll down to section *Visual Studio 2015, 2017 and 2019*.

For `lua51.dll`, I used this Microsoft download page:

* Microsoft Visual C++ 2015 Redistributable Update 3 (14.0.24212.0)
* https://www.microsoft.com/en-us/download/details.aspx?id=53587

The way everything works, like getting download sites from searches, it sets people up to download infected files, from sites that provide helpful information, but I guess that's not my problem.

**Comment by GezzMC** [comment]
Don't believe in leaving things open indefinitely. Anyone interested will have seen this. Web searches should find it. Thanks.

**Comment by pac-dev** [comment]
Thanks so much for figuring all this out. Obviously, I already have all those runtimes on my systems, so I didn't realize what was or wasn't required. As mentioned in #31, I'd like to provide all the DLLs in a portable way and will look into that.

**Comment by GezzMC** [comment]
Thanks for the reply. I reopen this issue rather than the other related one, since technically there's no problem other than a person needs to know what Visual C++ Redistributables to install. The other, [#31](https://github.com/pac-dev/protoplug/issues/31), might be of interest to someone not completely initiated in the constant problems of trying to get DAWs to work and keep working.

I reopen this because I did some more experimenting and found, at least for myself, that if I put `msvcp140.dll` and `vcruntime140.dll` in the path, and also use a more recent 64-bit `lua51.dll`, then Protoplug works without having to install any Visual C++ Redistributables.

* I got the Visual C++ Redistributables DLLs from a 64-bit Pale Moon browser distribution. The browser has updated itself, but originally it was`Palemoon-Portable-27.6.2.win64`, [palemoon.org/download.shtml](http://www.palemoon.org/download.shtml).
* The `lua51.dll` is from a 64-bit Terra distribution, `terra-Windows-x86_64-2e2032e.zip`, [github.com/zdevito/terra/releases](https://github.com/zdevito/terra/releases). The DLL file date is 2018-11-27.

This is a short summary. I'll post another comment with some more details and observations, like what I see other distributions doing with Visuall C++ Redistributables.

For myself, portability reigns supreme. I generally seek not to use programs that have magic installers that create mysterious or complicated dependencies, registry entries and that kind of thing.
